jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/FlowControlStrategyTest.java

```diff
@@ -554,95 +514,96 @@ public Stream.Listener onNewStream(Stream stream, HeadersFrame requestFrame)
+                streamRef2.set(stream);
             }
         });
+        await().atMost(5, TimeUnit.SECONDS).until(() -> streamRef2.get() != null);
 
         // Third request will consume the whole session window, which is now stalled.
-        // A fourth request will not be able to receive data.
-        List<Stream.Data> dataList3 = new ArrayList<>();
+        // A fourth request will not be able to receive data because the server is stalled.
+        AtomicReference<Stream> streamRef3 = new AtomicReference<>();
         MetaData.Request request3 = newRequest("GET", HttpFields.EMPTY);
         session.newStream(new HeadersFrame(request3, null, true), new Promise.Adapter<>(), new Stream.Listener()
         {
             @Override
             public void onDataAvailable(Stream stream)
             {
-                Stream.Data data = stream.readData();
-                if (!data.frame().isEndStream())
-                    stream.demand();
-                // Do not release it to stall flow control.
-                dataList3.add(data);
+                // Do not read to stall flow control.
+                streamRef3.set(stream);
             }
         });
+        await().atMost(5, TimeUnit.SECONDS).until(() -> streamRef3.get() != null);
 
         // Fourth request is now stalled.
-        CountDownLatch latch = new CountDownLatch(1);
+        AtomicReference<Stream> streamRef4 = new AtomicReference<>();
         MetaData.Request request4 = newRequest("GET", HttpFields.EMPTY);
         session.newStream(new HeadersFrame(request4, null, true), new Promise.Adapter<>(), new Stream.Listener()
         {
             @Override
             public void onDataAvailable(Stream stream)
             {
-                Stream.Data data = stream.readData();
-                data.release();
-                if (data.frame().isEndStream())
-                    latch.countDown();
-                else
-                    stream.demand();
+                streamRef4.set(stream);
             }
         });
-
         // Verify that the data does not arrive because the server session is stalled.
-        assertFalse(latch.await(1, TimeUnit.SECONDS));
+        await().during(1, TimeUnit.SECONDS).until(() -> streamRef4.get() == null);
 
         // Consume the data of the first response.
         // This will open up the session window, allowing the fourth stream to send data.
-        dataList1.forEach(Stream.Data::release);
+        consumeAll(streamRef1.get());
+        await().atMost(5, TimeUnit.SECONDS).until(() -> streamRef4.get() != null);
 
-        assertTrue(latch.await(5, TimeUnit.SECONDS));
+        consumeAll(streamRef2.get());
+        consumeAll(streamRef3.get());
+        consumeAll(streamRef4.get());
+    }
 
-        dataList2.forEach(Stream.Data::release);
-        dataList3.forEach(Stream.Data::release);
+    private void consumeAll(Stream stream) throws Exception
+    {
+        while (true)
```

Comment:
I would replace this loop with awaitility, like in `DataDemandTest` so that the test can never hang forever.